File: src/understudy/context.py
"""Render the event store into a compact transcript for the understudy model."""

from __future__ import annotations

from pathlib import Path

from understudy.events import Event, Kind
from understudy.store import EventStore


def render_activity(store: EventStore, *, max_events: int = 160, max_chars: int = 10000) -> str:
    """A token-frugal, most-recent-last view of the agent's activity."""
    lines = [_serialize(ev) for ev in store.events[-max_events:]]
    text = "\n".join(line for line in lines if line)
    if len(text) > max_chars:
        text = "…(earlier activity elided)…\n" + text[-max_chars:]
    return text or "(no activity yet)"


def _serialize(ev: Event) -> str:
    p = ev.payload
    t = ev.ts.strftime("%H:%M:%S")
    tag = " [subagent]" if ev.is_sidechain else ""
    match ev.kind:
        case Kind.USER_PROMPT:
            return f"{t} USER{tag}: {_clip(p.get('text', ''), 400)}"
        case Kind.ASSISTANT_TEXT:
            return f"{t} ASSISTANT{tag}: {_clip(p.get('text', ''), 400)}"
        case Kind.THINKING:
            text = p.get("text", "")
            body = _clip(text, 400) if text.strip() else "(not exposed)"
            return f"{t} THINKING{tag}: {body}"
        case Kind.TOOL_CALL:
            return f"{t} TOOL→ {p.get('name', '')}({_clip(_args(p), 160)}){tag}"
        case Kind.TOOL_RESULT:
            status = "ok" if p.get("ok", True) else "ERROR"
            return f"{t} TOOL← {p.get('name', '')} {status}: {_clip(p.get('summary', ''), 160)}{tag}"
        case Kind.FILE_EDIT:
            verb = "CREATE" if p.get("created") else "EDIT"
            name = Path(p.get("path", "")).name
            return f"{t} {verb} {name} +{p.get('added', 0)}-{p.get('removed', 0)}{tag}"
        case Kind.SESSION_START:
            return f"{t} SESSION cwd={p.get('cwd', '')}"
        case _:
            return ""


def _args(p: dict) -> str:
    inp = p.get("input") or {}
    for key in ("command", "file_path", "path", "pattern", "query", "url"):
        if key in inp and inp[key]:
            return f"{key}={inp[key]}"
    return ""


def _clip(s: str, n: int) -> str:
    s = " ".join(str(s).split())
    return s if len(s) <= n else s[: n - 1] + "…"

File: crates/core/src/context.rs
//! Render the event store into a compact transcript for the comprehension model.

use serde_json::Value;

use crate::events::{Event, EventKind};
use crate::store::{basename, EventStore};

pub fn render_activity(store: &EventStore, max_events: usize, max_chars: usize) -> String {
    let start = store.events.len().saturating_sub(max_events);
    let mut lines: Vec<String> = Vec::new();
    for ev in &store.events[start..] {
        let line = serialize(ev);
        if !line.is_empty() {
            lines.push(line);
        }
    }
    let mut text = lines.join("\n");
    if text.len() > max_chars {
        let mut cut = text.len() - max_chars;
        while cut < text.len() && !text.is_char_boundary(cut) {
            cut += 1;
        }
        text = format!("…(earlier activity elided)…\n{}", &text[cut..]);
    }
    if text.is_empty() {
        "(no activity yet)".to_string()
    } else {
        text
    }
}

/// One-line, most-recent-last representation of a single event (used by the
/// headless CLI and the activity feed).
pub fn event_line(ev: &Event) -> String {
    serialize(ev)
}

fn serialize(ev: &Event) -> String {
    let t = ev.ts.format("%H:%M:%S");
    let tag = if ev.is_sidechain { " [subagent]" } else { "" };
    match &ev.kind {
        EventKind::UserPrompt { text } => format!("{t} USER{tag}: {}", clip(text, 400)),
        EventKind::AssistantText { text } => format!("{t} ASSISTANT{tag}: {}", clip(text, 400)),
        EventKind::Thinking { text, .. } => {
            let body = if text.trim().is_empty() {
                "(not exposed)".to_string()
            } else {
                clip(text, 400)
            };
            format!("{t} THINKING{tag}: {body}")
        }
        EventKind::ToolCall { name, input, .. } => {
            format!("{t} TOOL→ {name}({}){tag}", clip(&args(input), 160))
        }
        EventKind::ToolResult { name, ok, summary, .. } => {
            let status = if *ok { "ok" } else { "ERROR" };
            format!("{t} TOOL← {name} {status}: {}{tag}", clip(summary, 160))
        }
        EventKind::FileEdit { path, added, removed, created, .. } => {
            let verb = if *created { "CREATE" } else { "EDIT" };
            format!("{t} {verb} {} +{added}-{removed}{tag}", basename(path))
        }
        EventKind::SessionStart { cwd, .. } => format!("{t} SESSION cwd={cwd}"),
        _ => String::new(),
    }
}

fn args(input: &Value) -> String {
    for key in ["command", "file_path", "path", "pattern", "query", "url"] {
        if let Some(v) = input.get(key) {
            let s = match v {
                Value::String(s) => s.clone(),
                Value::Null => String::new(),
                other => other.to_string(),
            };
            if !s.is_empty() {
                return format!("{key}={s}");
            }
        }
    }
    String::new()
}

/// Collapse whitespace and clip to `n` characters (char count, matching Python).
pub fn clip(s: &str, n: usize) -> String {
    let collapsed = s.split_whitespace().collect::<Vec<_>>().join(" ");
    if collapsed.chars().count() <= n {
        collapsed
    } else {
        let truncated: String = collapsed.chars().take(n.saturating_sub(1)).collect();
        format!("{truncated}…")
    }
}
